### PR TITLE
Add support for `textDocument/linkedEditingRange`

### LINKED ISSUE #987: Add support for linked editing
See https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_linkedEditingRange for more details.

One such feature could be to automatically rename the start and corresponding end tag if present.

### PR DESCRIPTION
**Pull request by angelozerr**
Add support for `textDocument/linkedEditingRange`

Fixes #987

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR requires LSP4J 0.111.0 which is not released (I'm using SNAPSHOT for the moment).

This PR requires the vscode-xml PR https://github.com/redhat-developer/vscode-xml/pull/426

File: pom.xml [hunk 1]
```diff
@@ -297,4 +297,17 @@
 			</build>
 		</profile>
 	</profiles>
+	<repositories>
+    <repository>
```
Comment (angelozerr) [inline pom.xml:301]: It should be removed when LSP4J 0.11.0 will be released

File: pom.xml [hunk 2]
```diff
@@ -7,7 +7,7 @@
 	<description>LemMinX is a XML Language Server Protocol (LSP), and can be used with any editor that supports LSP, to offer an outstanding XML editing experience</description>
 	<properties>
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-		<lsp4j.version>0.9.0</lsp4j.version>
+		<lsp4j.version>0.11.0-SNAPSHOT</lsp4j.version>
```
Comment (angelozerr) [inline pom.xml:10]: It should be replaced with `0.11.0` when LSP4J 0.11.0 will be released

**Comment by jonahgraham** [comment]
Thanks @angelozerr for commenting on https://github.com/eclipse/lsp4j/issues/466.

I have a question for you. How does this language server work with the Eclipse SimRel? Can it have a different version of LSP4J than the version that LSP4E uses and is part of SimRel? Is Lemminx part of SimRel (presumably via WWD if it is)? The answers to these questions will help determined when best to release 0.11.0 of LSP4J.

**Comment by jonahgraham** [comment]
PS Has lemminx had a release yet? None are listed on https://projects.eclipse.org/projects/technology.lemminx/governance but I thought that WWD was using it and it is making releases.

**Comment by rgrunber** [comment]
> I have a question for you. How does this language server work with the Eclipse SimRel? Can it have a different version of LSP4J than the version that LSP4E uses and is part of SimRel? Is Lemminx part of SimRel (presumably via WWD if it is)? The answers to these questions will help determined when best to release 0.11.0 of LSP4J.

Looking in WWD [snapshot builds](https://ci.eclipse.org/wildwebdeveloper/job/Wildwebdeveloper/job/master/336/consoleFull), they appear to be using lemminx from repo.eclipse.org and [embedding](https://github.com/eclipse/wildwebdeveloper/blob/master/org.eclipse.wildwebdeveloper.xml/pom.xml#L32) it. @angelozerr mentioned LemMinX will generally adopt the latest version of lsp4j so I'm guessing that's usually what lsp4e would use. I think as long as new changes can be broadcast to clients (like this one), we likely won't have issues updating.

> PS Has lemminx had a release yet? None are listed on https://projects.eclipse.org/projects/technology.lemminx/governance but I thought that WWD was using it and it is making releases.

I believe it's in incubation. The releases can be found at https://download.eclipse.org/lemminx/ .

**Comment by jonahgraham** [comment]
> > I have a question for you. How does this language server work with the Eclipse SimRel? Can it have a different version of LSP4J than the version that LSP4E uses and is part of SimRel? Is Lemminx part of SimRel (presumably via WWD if it is)? The answers to these questions will help determined when best to release 0.11.0 of LSP4J.
> 
> Looking in WWD [snapshot builds](https://ci.eclipse.org/wildwebdeveloper/job/Wildwebdeveloper/job/master/336/consoleFull), they appear to be using lemminx releases from repo.eclipse.org and [embedding](https://github.com/eclipse/wildwebdeveloper/blob/master/org.eclipse.wildwebdeveloper.xml/pom.xml#L32) it. @angelozerr mentioned LemMinX will generally adopt the latest version of lsp4j so I'm guessing that's usually what lsp4e would use. I think as long as new changes can be broadcast to clients (like this one), we likely won't have issues updating.

Thanks. It looks like wwd is using a lemminx uber jar, IIUC that means the version of lsp4j in lemminx can be different than that of lsp4e. We have pretty much decided not to release lsp4j 0.11.0 to 2021-03 simrel as we can't do it in time and let all adopters in simrel upgrade to that version. But if we do release 0.11.0 soon, then lemminx can start using it independently.

> 
> > PS Has lemminx had a release yet? None are listed on https://projects.eclipse.org/projects/technology.lemminx/governance but I thought that WWD was using it and it is making releases.
> 
> I believe it's in incubation. The releases can be found at https://download.eclipse.org/lemminx/ .

OK. I think it still needs release records for the releases and a release review once per year. Being in incubation does not matter for that requirement of EDP.

**Comment by angelozerr** [comment]
> Thanks. It looks like wwd is using a lemminx uber jar, IIUC that means the version of lsp4j in lemminx can be different than that of lsp4e. 

Exactly and my draft PR which consumes 0.11.0 works with the current LSP4E which doesn't consume 0.11.0.

> We have pretty much decided not to release lsp4j 0.11.0 to 2021-03 simrel as we can't do it in time and let all adopters in simrel upgrade to that version. But if we do release 0.11.0 soon, then lemminx can start using it independently.

If I understand correctly you mean that you could do a release next week for instance? If it that it should be really fantastic because we have the intention to do a release of vscode-xml soon and we could provide this linked editing support inside vscode-xml.

**Comment by jonahgraham** [comment]
> > We have pretty much decided not to release lsp4j 0.11.0 to 2021-03 simrel as we can't do it in time and let all adopters in simrel upgrade to that version. But if we do release 0.11.0 soon, then lemminx can start using it independently.
> 
> If I understand correctly you mean that you could do a release next week for instance? If it that it should be really fantastic because we have the intention to do a release of vscode-xml soon and we could provide this linked editing support inside vscode-xml.

I am still waiting (gently pushing) to get LSP 3.16.0 support finished in LSP4J 0.11.0. I only meant that we can release 0.11.0 independently of the release train and that you could start using it, also independent of the release train. Sorry for not giving you more/better visibility on what it will take for 0.11.0 to release. https://github.com/eclipse/lsp4j/milestone/18 has the progress on getting all the issues resolved.

**Comment by rgrunber** [comment]
FWIW, we'll likely end up sticking with lsp4j 0.9.0 for our upcoming release, and update to latest afterwards, so this isn't blocking us.

**Comment by jonahgraham** [comment]
> FWIW, we'll likely end up sticking with lsp4j 0.9.0 for our upcoming release, and update to latest afterwards, so this isn't blocking us.

Thanks for letting me know - the LSP4J release is imminent now - https://github.com/eclipse/lsp4j/issues/466#issuecomment-793237792

**Comment by jonahgraham** [comment]
The LSP4J release is now done: https://github.com/eclipse/lsp4j/issues/466#issuecomment-802895834

File: pom.xml (hunk 1 above)
Comment (angelozerr) [reply pom.xml]: fixed

File: pom.xml (hunk 2 above)
Comment (angelozerr) [reply pom.xml]: fixed

**Comment by angelozerr** [comment]
> The LSP4J release is now done: eclipse/lsp4j#466 (comment)

Thanks @jonahgraham for your feedback. I have updated my PR to consume now the 0.11.0 release of LSP4J.

**Comment by datho7561** [comment]
When using lsp4j 0.12.0, this PR breaks the server. The server will issue diagnostics when it first starts, but when the other capabilities are dynamically registered, the server becomes unresponsive.

If I change the lsp4j version to 0.11.0, the server and the linked editing works great.

**Comment by fbricon** [comment]
Binary builds fail with 
```
Error: Error parsing reflection configuration in jar:file:/home/runner/work/lemminx/lemminx/org.eclipse.lemminx/target/org.eclipse.lemminx-0.16.1-SNAPSHOT.jar!/META-INF/native-image/reflect-config.json:
Could not resolve org.eclipse.lsp4j.SemanticHighlightingCapabilities for reflection configuration. To allow unresolvable reflection configuration, use option -H:+AllowIncompleteClasspath
Verify that the configuration matches the schema described in the -H:PrintFlags=+ output for option ReflectionConfigurationResources.
Error: Use -H:+ReportExceptionStackTraces to print stacktrace of underlying exception

```

**Comment by datho7561** [comment]
Binary still doesn't work from my testing

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMElement.java [hunk 3]
```diff
@@ -287,11 +287,15 @@ public boolean isInStartTag(int offset) {
...
+	}
+
+	public boolean isInEndTag(int offset, boolean afterBackSlash) {
 		if (endTagOpenOffset == NULL_VALUE) {
 			// case >|
 			return false;
 		}
-		if (offset > endTagOpenOffset && offset < getEnd()) {
+		if (offset > endTagOpenOffset + (afterBackSlash ? 1 : 0) && offset < getEnd()) {
 			// case <\bean | >
```
Comment (datho7561) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMElement.java:299]: `</bean | >`

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsi/XSISchemaModel.java [hunk 4]
```diff
@@ -159,36 +155,42 @@ private static void createCompletionItem(String attrName, boolean canSupportSnip
...
 		if (attrAtOffset == null) {
 			return;
 		}
 		String attrName = attrAtOffset.getName();
-		
-		if(attrName != null) {
-			if(actualPrefix != null && attrName.equals(actualPrefix + ":nil")) { // Value completion for 'nil' attribute
+
+		if (attrName != null) {
+			if (actualPrefix != null && attrName.equals(actualPrefix + ":nil")) { // Value completion for 'nil'
```
Comment (datho7561) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsi/XSISchemaModel.java:172]: Put the comment on the line before the `if` statement

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsi/XSISchemaModel.java [hunk 5]
```diff
@@ -159,36 +155,42 @@ private static void createCompletionItem(String attrName, boolean canSupportSnip
...
+																					// attribute
 				createCompletionItemsForValues(StringUtils.TRUE_FALSE_ARRAY, document, request, response);
-			}
-			else if(document.getDocumentElement() != null && document.getDocumentElement().equals(nodeAtOffset)) { // if in the root element
-				if(attrName.equals("xmlns:xsi")) {
+			} else if (document.getDocumentElement() != null && document.getDocumentElement().equals(nodeAtOffset)) { // if
```
Comment (datho7561) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsi/XSISchemaModel.java:175]: Put the comment on the line before the `if` statement

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLLinkedEditing.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * XML linked editing support.
+ *
+ */
+class XMLLinkedEditing {
+
+	private static Logger LOGGER = Logger.getLogger(XMLLinkedEditing.class.getName());
+
+	public LinkedEditingRanges findLinkedEditingRanges(DOMDocument document, Position position,
```
Comment (datho7561) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLLinkedEditing.java:37]: Add javadoc

File: pom.xml [hunk 7]
```diff
@@ -4,11 +4,10 @@
 	<artifactId>lemminx-parent</artifactId>
 	<version>0.16.1-SNAPSHOT</version>
 	<packaging>pom</packaging>
-	<name>Eclipse LemMinX</name>
```
Comment (datho7561) [inline pom.xml:7]: Do not remove project name

**Review by datho7561 — CHANGES_REQUESTED** [review]

**Comment by datho7561** [comment]
I found that this case is broken:

mirror-cursor-glitch

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMElement.java (hunk 3 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMElement.java:299]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsi/XSISchemaModel.java (hunk 4 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsi/XSISchemaModel.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsi/XSISchemaModel.java (hunk 5 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsi/XSISchemaModel.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLLinkedEditing.java (hunk 6 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLLinkedEditing.java:47]: fixed

File: pom.xml (hunk 7 above)
Comment (angelozerr) [reply pom.xml:7]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMElement.java (hunk 3 above)
Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMElement.java:299]: Hey Angelo, I think you got the wrong line, I was talking about the comment on line 299

File: pom.xml (hunk 7 above)
Comment (datho7561) [reply pom.xml:7]: I think the name should directly be under `<project>`, instead of being under `<properties>`.

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMElement.java (hunk 3 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMElement.java]: my bad,sorry-(

File: pom.xml (hunk 7 above)
Comment (angelozerr) [reply pom.xml]: my bad,sorry-( fixed

**Comment by angelozerr** [comment]
> I found that this case is broken:

fixed.

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLLinkedEditingTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+		// testLinkedEditingFor("<div><div|</div>", null);
+		// testLinkedEditingFor("<div><div><div|</div></div>", null);
+
+		testLinkedEditingFor("<div| ></div>", le(r(0, 1, 0, 4), r(0, 8, 0, 11)));
+		testLinkedEditingFor("<div| id='foo'></div>", le(r(0, 1, 0, 4), r(0, 16, 0, 19)));
+
+		// FIXME
+		// testLinkedEditingFor("<|></>", le(r(0, 1, 0, 1), r(0, 4, 0, 4)));
+		// testLinkedEditingFor("<><div></div></|>", le(r(0, 1, 0, 1), r(0, 15, 0,
+		// 15)));
```
Comment (datho7561) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLLinkedEditingTest.java:63]: These tests are still commented out. I think it's worthwhile to run these tests.

File: org.eclipse.lemminx/src/main/resources/META-INF/native-image/reflect-config.json [hunk 9]
```diff
@@ -1122,6 +1122,46 @@
 			"name": "value"
 		}]
 	},
+	{
```
Comment (datho7561) [inline org.eclipse.lemminx/src/main/resources/META-INF/native-image/reflect-config.json:1125]: The binary server doesn't work

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLLinkedEditingTest.java (hunk 8 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLLinkedEditingTest.java]: I have adapted tests.

File: org.eclipse.lemminx/src/main/resources/META-INF/native-image/reflect-config.json (hunk 9 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/resources/META-INF/native-image/reflect-config.json:1183]: Please retry.

File: org.eclipse.lemminx/src/main/resources/META-INF/native-image/reflect-config.json [hunk 10]
```diff
@@ -1118,8 +1166,58 @@
 	},
 	{
 		"name": "org.eclipse.lsp4j.InsertTextFormat",
-		"fields": [{
-			"name": "value"
+		"allDeclaredFields": true,
+		"methods": [{
+			"name": "<init>",
+			"parameterTypes": ["int"]
+		}]
```
Comment (datho7561) [inline org.eclipse.lemminx/src/main/resources/META-INF/native-image/reflect-config.json:1173]: Replace with:
```json
"allDeclaredMethods": true
```

**Review by datho7561 — CHANGES_REQUESTED** [review]

**Review by datho7561 — COMMENTED** [review]
A few more reflection configuration changes that need to be made:

Modify:
```diff
        {
                "name": "org.eclipse.lsp4j.InsertTextMode",
-               "fields": [{
-                       "name": "value"
-               }]
+               "allDeclaredFields": true
        },
```

Then, add:
```json
{
         "name": "org.eclipse.lsp4j.SetTraceParams",
         "allDeclaredFields": true,
         "methods": [{
                  "name": "<init>",
                "parameterTypes": []
        }]
},
```

File: org.eclipse.lemminx/src/main/resources/META-INF/native-image/reflect-config.json (hunk 10 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/resources/META-INF/native-image/reflect-config.json]: fixed

**Comment by angelozerr** [comment]
All should be fixed, thanks for your feedback.

**Comment by datho7561** [comment]
One more reflection entry to add (sorry, I think this is the last):

```json
{
        "name": "org.eclipse.lsp4j.adapters.CompletionItemTextEditTypeAdapter",
        "methods": [{
        "name": "<init>",
                 "parameterTypes": []
        }]
},
```

**Review by datho7561 — APPROVED** [review]

**Pull request merged**